### PR TITLE
Banners

### PR DESCRIPTION
**Pull request by ZephireNZ**
This adds Banners to Glowstone. It also requires https://github.com/GlowstoneMC/Glowkit/pull/24 to build. It also includes the changes from #334, which is why there is changes to GlowSign.

TileEntities, the BlockState, ItemMeta, etc are all implemented, and seemed to work when testing.

### CODE REVIEW

**Comment by ZephireNZ** [comment]
Related #119

**Comment by gdude2002** [comment]
Just to confirm - with [Glowkit PR 24](https://github.com/GlowstoneMC/Glowkit/pull/24), this does indeed appear to build perfectly. http://bamboo.gserv.me/browse/GLOW-SRV-PR360-47

**Comment by gdude2002** [comment]
Broken as of the Lombok commit. http://bamboo.gserv.me/browse/GLOW-SRV-PR360-49

**Comment by gdude2002** [comment]
Fixed: http://bamboo.gserv.me/browse/GLOW-SRV-PR360-51

**Comment by turt2live** [comment]
@ZephireNZ Your PR is currently not mergable. Please resolve the merge conflicts as soon as possible, thanks!

`PR_BOT_CODE:1` (you can ignore this)

**Comment by dequis** [comment]
Trivial merge conflict in the imports part of GlowPlayer

Merged version: https://github.com/dequis/Glowstone/compare/pr-360

**Comment by turt2live** [comment]
Note: saving errors occur with the banners.

Test plugin used: [Google Drive](https://drive.google.com/folderview?id=0B6JIYSBkLkAnd1JicnVXTU9GUHc&usp=sharing&tid=0B6JIYSBkLkAnTTZYd3Z4ZTMzV3c)

**Comment by ZephireNZ** [comment]
Yey, it builds: http://bamboo.gserv.me/browse/GLOW-SRV-PR360-85

**Comment by bendem** [comment]
Hm, when trying to craft a banner, it throws some

```
12:56:56 [GRAVE] Error in network input
io.netty.handler.codec.DecoderException: java.lang.IllegalArgumentException: Compound does not contain ByteTag "Color"
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:163)
        at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:108)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:787)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:125)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: Compound does not contain ByteTag "Color"
        at net.glowstone.util.nbt.CompoundTag.getTag(CompoundTag.java:266)
        at net.glowstone.util.nbt.CompoundTag.get(CompoundTag.java:260)
        at net.glowstone.util.nbt.CompoundTag.getByte(CompoundTag.java:67)
        at net.glowstone.block.blocktype.BlockBanner.fromNBT(BlockBanner.java:86)
        at net.glowstone.inventory.GlowMetaBanner.readNbt(GlowMetaBanner.java:54)
        at net.glowstone.inventory.GlowItemFactory.readNbt(GlowItemFactory.java:79)
        at net.glowstone.net.GlowBufUtils.readSlot(GlowBufUtils.java:175)
        at net.glowstone.net.codec.play.inv.WindowClickCodec.decode(WindowClickCodec.java:19)
        at net.glowstone.net.codec.play.inv.WindowClickCodec.decode(WindowClickCodec.java:11)
        at net.glowstone.net.pipeline.CodecsHandler.decode(CodecsHandler.java:53)
        at net.glowstone.net.pipeline.CodecsHandler.decode(CodecsHandler.java:19)
        at io.netty.handler.codec.MessageToMessageCodec$2.decode(MessageToMessageCodec.java:81)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
        ... 24 more
```

EDIT: Also, when breaking the block a banner is attached to, it doesn't pop off the banner.

File: src/main/java/net/glowstone/entity/GlowPlayer.java [hunk 1]
```diff
@@ -1360,6 +1362,18 @@ public void sendSignChange(Location location, String[] lines) throws IllegalArgu
         afterBlockChanges.add(UpdateSignMessage.fromPlainText(location.getBlockX(), location.getBlockY(), location.getBlockZ(), lines));
     }
 
+    public void sendBlockEntityChange(Location location, CompoundTag nbt, GlowBlockEntity type) throws NullPointerException {
```
Comment (turt2live) [inline src/main/java/net/glowstone/entity/GlowPlayer.java]: throws is not required

File: src/main/java/net/glowstone/entity/GlowPlayer.java [hunk 2]
```diff
@@ -1360,6 +1362,18 @@ public void sendSignChange(Location location, String[] lines) throws IllegalArgu
...
+        Validate.notNull(location, "Location cannot be null");
+        Validate.notNull(nbt, "NBT cannot be null");
+        Validate.notNull(type, "Type cannot be null");
+
+        afterBlockChanges.add(new UpdateBlockEntityMessage(location.getBlockX(), location.getBlockY(), location.getBlockZ(), type.getValue(), nbt));
+    }
+
+    public void sendBannerChange(Location location, CompoundTag nbt) throws NullPointerException {
```
Comment (turt2live) [inline src/main/java/net/glowstone/entity/GlowPlayer.java]: throws is not required

**Comment by turt2live** [comment]
This looks ready to pull despite the extra `throws` statements. Although Magic values are in Bukkit, I don't really think it's worth being picky about it right now considering the whole situation in Bukkit is a bit strange with respect to materials.

**Comment by turt2live** [comment]
@ZephireNZ Due to recent changes on master, this PR is no longer mergable. Please rebase as soon as possible!

**Comment by dequis** [comment]
Easy conflict on ItemTable

Merged branch here https://github.com/dequis/Glowstone/compare/pr-360

@SpaceManiac could consider it still ready to pull, tbh.

**Comment by gdude2002** [comment]
Failing: http://bamboo.gserv.me/browse/GLOW-SRV-PR360-91

**Comment by ZephireNZ** [comment]
@turt2live Moving back to WIP, made some blunders with the way I've done things. Specifically, you can't add duplicate layers (due to using a Map) like you can in vanilla.

**Comment by ZephireNZ** [comment]
OK, that was quick haha. Fixed it now, feel free to update your plugin and test again.

**Comment by turt2live** [comment]
@ZephireNZ banner background colors are not saved correctly and default to white on server restart.

**Comment by dequis** [comment]
Simple conflicts in ItemTable:
- `import *` - keep from master
- Added items - keep both sides

Merged branch here: https://github.com/dequis/Glowstone/compare/pr-360

(totally not copypasted from the previous merged PR)

**Comment by ZephireNZ** [comment]
@turt2live I believe your bot derped, still needs changes :P

**Comment by turt2live** [comment]
It assumes that since you committed something that you fixed it.

sent from mobile

**Comment by turt2live** [comment]
@ZephireNZ Is this ready to test again? Your last comment indicated changes were required.

**Comment by ZephireNZ** [comment]
Yes, it should be good to go. Turns out the problem was pretty trivial (checking for a byte, when it's an int value).

**Comment by ZephireNZ** [comment]
@turt2live Just a rebase, shouldn't even touch any affected areas. I understand if you want to retest, but...

**Comment by turt2live** [comment]
@SpaceManiac @ZephireNZ Beware of the code duplication in #267, `GlowPlayer#sendBlockEntityChange()` when this (or #267) is pulled.

**Comment by SpaceManiac** [comment]
Merged in 05dbe066cfbb0ecfd28225204161a00c961e4877.

**Pull request closed without merging**